presto-main/src/main/java/io/prestosql/operator/PartitionedOutputOperator.java

```diff
@@ -261,8 +262,13 @@ public boolean isFinished()
     @Override
     public ListenableFuture<?> isBlocked()
     {
-        ListenableFuture<?> blocked = partitionFunction.isFull();
-        return blocked.isDone() ? NOT_BLOCKED : blocked;
+        if (isBlocked.isDone()) {
+            isBlocked = partitionFunction.isFull();
```

Comment:
future for `partitionFunction` is already cached

Reply:
If you mean the cached future in `OutputBufferMemoryManager` then you're technically correct, except that retrieving that instance is through `OutputBufferMemoryManager#getBufferBlockedFuture()` which needs to synchronize on the heavily lock on each access.

This is especially painful since each thread working through the same task's driver loop necessarily shares the output buffer instance and will each call 3 different methods in rapid succession that all check whether the operator is currently blocked (`isBlocked()`, `needsInput()`, `isFinished()`). This is on top of:
- HTTP threads decrementing memory when acknowledging pages
- other worker threads incrementing memory after enqueuing pages
- Memory tracker future callbacks
- Constant checks against `OutputBufferMemoryManager#isOverutilized()` task concurrency levels.
- The fact that [this call into the system memory context](https://github.com/prestosql/presto/blob/37c74ad15f5f0c9c770c2aa491e49ffe5146d4b1/presto-main/src/main/java/io/prestosql/execution/buffer/OutputBufferMemoryManager.java#L91) is in effect entering a single global lock tree for every memory tracker update in any operator on any thread anywhere, which means that a thread holding the output buffer monitor can and does often block trying to acquire the memory tracking lock which subsequently locks out all threads calling any synchronized method on the output buffer.

This change _somewhat_ reduces the pain of all of that by caching the future in the operator itself when blocked so that subsequent checks don't re-enter the same massive bottleneck, but offers no improvement when not blocked. I'm in the process of putting together a proprosal to address the memory tracking and output buffer issues now.

Reply:
Please add a comment in code before `if (isBlocked.isDone()) {`

Reply:
Added